protocol/decoder/decoder.go

```diff
@@ -0,0 +1,271 @@
+		return -1, errors.NewPacketDecodingError(errorMessage).AddContexts("INT16", variableName)
+	}
+
+	decodedInteger := int16(binary.BigEndian.Uint16(d.bytes[d.offset:]))
+	d.offset += 2
+	d.logDecodedValue(variableName, decodedInteger)
+	return decodedInteger, nil
+}
+
+func (d *Decoder) ReadInt32(variableName string) (int32, error) {
+	if d.UnreadBytesCount() < 4 {
+		rem := d.UnreadBytesCount()
+		errorMessage := fmt.Sprintf("Expected int32 length to be 4 bytes, got %d bytes", rem)
+		err := errors.NewPacketDecodingError(errorMessage).AddContexts("INT32", variableName)
+		return -1, err
+	}
+
+	decodedInteger := int32(binary.BigEndian.Uint32(d.bytes[d.offset:]))
+	d.offset += 4
+	d.logDecodedValue(variableName, decodedInteger)
+	return decodedInteger, nil
+}
+
+func (d *Decoder) ReadUnsignedVarint(variableName string) (uint64, error) {
+	decodedInteger, n := binary.Uvarint(d.bytes[d.offset:])
+
+	if n == 0 {
+		return 0, errors.NewPacketDecodingError("Unexpected end of data").AddContexts("UNSIGNED_VARINT", variableName)
+	}
+
+	if n < 0 {
+		errorMessage := fmt.Sprintf("Unexpected unsigned varint overflow after decoding %d bytes", -n)
+		return 0, errors.NewPacketDecodingError(errorMessage).AddContexts("UNSIGNED_VARINT", variableName)
+	}
+
+	d.offset += n
+	d.logDecodedValue(variableName, decodedInteger)
+	return decodedInteger, nil
+}
+
+// Composite Types : Uses Basic Types for decoding
+
+func (d *Decoder) ReadArrayLength(variableName string) (int, error) {
+	if d.UnreadBytesCount() < 4 {
+		rem := d.UnreadBytesCount()
+		errorMessage := fmt.Sprintf("Expected array length prefix to be 4 bytes, got %d bytes", rem)
+		return -1, errors.NewPacketDecodingError(errorMessage).AddContexts("ARRAY_LENGTH", variableName)
+	}
+	decodedInteger, err := d.ReadInt32("")
+
+	if err != nil {
+		if decodingErr, ok := err.(*errors.PacketDecodingError); ok {
+			return -1, decodingErr.AddContexts("ARRAY_LENGTH", variableName)
+		}
+		return -1, err
+	}
+
+	arrayLength := int(decodedInteger)
+
+	if arrayLength > d.UnreadBytesCount() {
+		errorMessage := fmt.Sprintf("Expect to read at least %d bytes, but only %d bytes are remaining", arrayLength, d.UnreadBytesCount())
+		return -1, errors.NewPacketDecodingError(errorMessage).AddContexts("ARRAY_LENGTH", variableName)
+	} else if (arrayLength > 2*math.MaxUint16) || (arrayLength < 0) {
+		return -1, errors.NewPacketDecodingError(fmt.Sprintf("Invalid array length: %d", arrayLength)).AddContexts("ARRAY_LENGTH", variableName)
+	}
+
+	d.logDecodedValue(variableName, arrayLength)
+	return arrayLength, nil
+}
+
+func (d *Decoder) ReadCompactArrayLength(variableName string) (int, error) {
+	decodedInteger, err := d.ReadUnsignedVarint("")
+
+	if err != nil {
+		if decodingErr, ok := err.(*errors.PacketDecodingError); ok {
+			return 0, decodingErr.AddContexts("COMPACT_ARRAY_LENGTH", variableName)
+		}
+		return 0, err
+	}
+
+	if decodedInteger == 0 {
```

Comment:
Is it intended that here we read a value and don't log it? Also, we're essentially treating the value 1 and 0 the same (here we return 0 directly, below we do `1-1` which is also 0)